src/data/iterative_device_dmatrix.cu

```diff
@@ -0,0 +1,109 @@
+/*!
+ * Copyright 2020 XGBoost contributors
+ */
+#include <memory>
+#include <type_traits>
+
+#include "../common/quantile.cuh"
+#include "iterative_device_dmatrix.h"
+#include "sparse_page_source.h"
+#include "ellpack_page.cuh"
+#include "proxy_dmatrix.h"
+#include "device_adapter.cuh"
+
+namespace xgboost {
+namespace data {
+
+#define DISPATCH_MEM(__Proxy, __Fn)                                     \
+  [](DMatrixProxy const* proxy) -> decltype(                            \
+      (dmlc::get<CupyAdapterBatch>(proxy->Value())).__Fn()) {           \
+    if (proxy->Value().type() == typeid(CupyAdapterBatch)) {            \
```

Comment:
Is it correct to say that the proxy matrix performs type erasure for the batch, and now we're trying to recover the type of the batch dynamically?

From the way I see it, we have an implicit list of allowable types, namely `CupyAdapterBatch` and `CudfAdapterBatch`. It may be better to add a type ID field in the DMatrix proxy. [Take a look at PackedFunc](https://docs.tvm.ai/dev/runtime.html#packedfunc) from TVM project, which uses type erasure to expose a generic function type. The PackedFunc uses type ID to distinguish between underlying types.

Also, is it possible to avoid macro here? Why use macro?

Reply:
I will look into the pack function today.

Reply:
Cool, I felt that PackedFunc may give us hint for simplifying the proxy.